File: python/geny_harness/session/freshness.py
"""Session freshness policy."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from enum import Enum


class FreshnessStatus(str, Enum):
    FRESH = "fresh"
    STALE_WARN = "stale_warn"
    STALE_IDLE = "stale_idle"
    STALE_COMPACT = "stale_compact"
    STALE_RESET = "stale_reset"

    @property
    def should_revive(self) -> bool:
        return self == FreshnessStatus.STALE_IDLE

    @property
    def should_compact(self) -> bool:
        return self == FreshnessStatus.STALE_COMPACT

    @property
    def should_reset(self) -> bool:
        return self == FreshnessStatus.STALE_RESET


class FreshnessPolicy:
    def __init__(
        self,
        idle_timeout: timedelta = timedelta(minutes=30),
        warn_threshold: timedelta = timedelta(minutes=20),
        compact_message_count: int = 100,
        reset_message_count: int = 500,
        max_age: timedelta = timedelta(hours=4),
    ):
        self.idle_timeout = idle_timeout
        self.warn_threshold = warn_threshold
        self.compact_message_count = compact_message_count
        self.reset_message_count = reset_message_count
        self.max_age = max_age

    def evaluate(
        self, created_at: datetime, last_active: datetime, message_count: int
    ) -> FreshnessStatus:
        now = datetime.now(timezone.utc)
        age = now - created_at
        idle = now - last_active

        if age > self.max_age:
            return FreshnessStatus.STALE_RESET
        if message_count >= self.reset_message_count:
            return FreshnessStatus.STALE_RESET
        if message_count >= self.compact_message_count:
            return FreshnessStatus.STALE_COMPACT
        if idle > self.idle_timeout:
            return FreshnessStatus.STALE_IDLE
        if idle > self.warn_threshold:
            return FreshnessStatus.STALE_WARN
        return FreshnessStatus.FRESH

File: crates/geny-harness-core/src/session/freshness.rs
//! Session health evaluation.

use chrono::{DateTime, Duration, Utc};
use serde::{Deserialize, Serialize};

/// Session freshness status.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum FreshnessStatus {
    /// Normal operation.
    Fresh,
    /// Idle timeout approaching.
    StaleWarn,
    /// Idle for extended period.
    StaleIdle,
    /// Many messages accumulated.
    StaleCompact,
    /// Max age or message count exceeded.
    StaleReset,
}

impl FreshnessStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            FreshnessStatus::Fresh => "fresh",
            FreshnessStatus::StaleWarn => "stale_warn",
            FreshnessStatus::StaleIdle => "stale_idle",
            FreshnessStatus::StaleCompact => "stale_compact",
            FreshnessStatus::StaleReset => "stale_reset",
        }
    }

    pub fn should_revive(&self) -> bool {
        matches!(self, FreshnessStatus::StaleIdle)
    }

    pub fn should_compact(&self) -> bool {
        matches!(self, FreshnessStatus::StaleCompact)
    }

    pub fn should_reset(&self) -> bool {
        matches!(self, FreshnessStatus::StaleReset)
    }
}

/// Session freshness thresholds.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FreshnessPolicy {
    pub idle_timeout_secs: i64,
    pub warn_threshold_secs: i64,
    pub compact_message_count: usize,
    pub reset_message_count: usize,
    pub max_age_secs: i64,
}

impl Default for FreshnessPolicy {
    fn default() -> Self {
        Self {
            idle_timeout_secs: 30 * 60,      // 30 minutes
            warn_threshold_secs: 20 * 60,    // 20 minutes
            compact_message_count: 100,
            reset_message_count: 500,
            max_age_secs: 4 * 60 * 60,       // 4 hours
        }
    }
}

impl FreshnessPolicy {
    pub fn evaluate(
        &self,
        created_at: DateTime<Utc>,
        last_active: DateTime<Utc>,
        message_count: usize,
    ) -> FreshnessStatus {
        let now = Utc::now();
        let age = now - created_at;
        let idle = now - last_active;

        // Priority: max_age > reset_message_count > compact_message_count > idle > warn > FRESH
        if age > Duration::seconds(self.max_age_secs) {
            return FreshnessStatus::StaleReset;
        }
        if message_count >= self.reset_message_count {
            return FreshnessStatus::StaleReset;
        }
        if message_count >= self.compact_message_count {
            return FreshnessStatus::StaleCompact;
        }
        if idle > Duration::seconds(self.idle_timeout_secs) {
            return FreshnessStatus::StaleIdle;
        }
        if idle > Duration::seconds(self.warn_threshold_secs) {
            return FreshnessStatus::StaleWarn;
        }
        FreshnessStatus::Fresh
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_freshness_defaults() {
        let policy = FreshnessPolicy::default();
        let now = Utc::now();
        assert_eq!(policy.evaluate(now, now, 0), FreshnessStatus::Fresh);
    }

    #[test]
    fn test_freshness_stale_reset_message_count() {
        let policy = FreshnessPolicy::default();
        let now = Utc::now();
        assert_eq!(
            policy.evaluate(now, now, 500),
            FreshnessStatus::StaleReset
        );
    }

    #[test]
    fn test_freshness_stale_compact() {
        let policy = FreshnessPolicy::default();
        let now = Utc::now();
        assert_eq!(
            policy.evaluate(now, now, 100),
            FreshnessStatus::StaleCompact
        );
    }

    #[test]
    fn test_freshness_status_properties() {
        assert!(FreshnessStatus::StaleIdle.should_revive());
        assert!(!FreshnessStatus::Fresh.should_revive());
        assert!(FreshnessStatus::StaleCompact.should_compact());
        assert!(FreshnessStatus::StaleReset.should_reset());
    }
}
